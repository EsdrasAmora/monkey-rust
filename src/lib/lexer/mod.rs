mod token;

pub fn print_token() {
    println!("Hello world, {:?}", token::Token::Value);
}
