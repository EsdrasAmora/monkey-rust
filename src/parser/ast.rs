#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Statement {
    Let {
        identifier: String,
        value: Box<Expression>,
    },
    Return(Box<Expression>),
    Expression(Box<Expression>),
}

//TODO: create a builder for this.
#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Expression {
    Literal(Literal),
    Identifier(String),
    Oposite(Box<Expression>),
    Not(Box<Expression>),
    Eq(BynaryExpression),
    NotEq(BynaryExpression),
    LT(BynaryExpression),
    GT(BynaryExpression),
    Add(BynaryExpression),
    Sub(BynaryExpression),
    Mul(BynaryExpression),
    Div(BynaryExpression),
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct BynaryExpression {
    lhs: Box<Expression>,
    rhs: Box<Expression>,
}
//maybe precedence should be a method on Token.
impl Expression {
    fn boxed(self) -> Box<Self> {
        Box::new(self)
    }

    fn precedence(&self) -> u8 {
        match self {
            Expression::Literal(_) | Expression::Identifier(_) => 0,
            Expression::Oposite(_) | Expression::Not(_) => 1,
            Expression::Eq(_) | Expression::NotEq(_) => 2,
            Expression::LT(_) | Expression::GT(_) => 3,
            Expression::Mul(_) | Expression::Div(_) => 4,
            Expression::Add(_) | Expression::Sub(_) => 5,
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Literal {
    Int(i64),
    String(String),
    True,
    False,
    Nill,
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Expression::Literal(literal)
    }
}
