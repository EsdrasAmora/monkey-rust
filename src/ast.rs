use crate::token::{Identifier, Token};
use serde::Serialize;
use smol_str::SmolStr;

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum Statement {
    Let {
        identifier: Identifier,
        value: Box<Expression>,
    },
    Return(Box<Expression>),
    Expression(Box<Expression>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum Expression {
    Literal(Literal),
    Identifier(Identifier),
    BinaryExp(BinaryExpression),
    UnaryExpression(UnaryExpression),
    If(IfExpression),
    IndexExpression(IndexExpression),
    Function(FunctionExpression),
    Call(CallExpression),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum BinaryOperator {
    Eq,
    NotEq,
    Lt,
    Lte,
    Gt,
    Gte,
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum UnaryOperator {
    Not,
    Minus,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct UnaryExpression {
    pub value: Box<Expression>,
    pub operator: UnaryOperator,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct CallExpression {
    pub arguments: Vec<Expression>,
    pub function: Box<Expression>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct IndexExpression {
    pub container: Box<Expression>,
    pub index: Box<Expression>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct FunctionExpression {
    pub parameters: Vec<Identifier>,
    pub body: BlockStatement,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct IfExpression {
    pub condition: Box<Expression>,
    pub consequence: BlockStatement,
    pub alternative: Option<BlockStatement>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct BinaryExpression {
    pub operator: BinaryOperator,
    pub lhs: Box<Expression>,
    pub rhs: Box<Expression>,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Default)]
pub struct BlockStatement(pub Vec<Statement>);

impl BlockStatement {
    pub fn new(value: Vec<Statement>) -> Self {
        Self(value)
    }
}

impl Expression {
    pub fn boxed(self) -> Box<Self> {
        Box::new(self)
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub enum Literal {
    Int(i64),
    String(SmolStr),
    Array(Vec<Expression>),
    Hash(Vec<(Expression, Expression)>),
    True,
    False,
    Nil,
}

impl Literal {
    pub fn into_exp(self) -> Box<Expression> {
        Expression::Literal(self).boxed()
    }
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Expression::Literal(literal)
    }
}

impl Token {
    #[inline]
    pub fn binary_expression_type(&self) -> Option<BinaryOperator> {
        match self {
            Token::Plus => Some(BinaryOperator::Add),
            Token::Minus => Some(BinaryOperator::Sub),
            Token::Slash => Some(BinaryOperator::Div),
            Token::Asterisk => Some(BinaryOperator::Mul),
            Token::Eq => Some(BinaryOperator::Eq),
            Token::NotEq => Some(BinaryOperator::NotEq),
            Token::Lt => Some(BinaryOperator::Lt),
            Token::Lte => Some(BinaryOperator::Lte),
            Token::Gt => Some(BinaryOperator::Gt),
            Token::Gte => Some(BinaryOperator::Gte),
            _ => None,
        }
    }
}
