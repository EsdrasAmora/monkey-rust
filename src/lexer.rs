use crate::token::{Identifier, Token};
use smol_str::SmolStr;
use std::iter::{self, Peekable};

#[derive(Debug)]
pub struct Lexer {
    pub tokens: Vec<Token>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        let mut chars = input.trim().chars().filter(|x| x.is_ascii()).peekable();
        let mut tokens = Vec::with_capacity(32);

        while let Some(char) = chars.next() {
            if char.is_whitespace() {
                continue;
            }
            let token = Lexer::new_helper(char, &mut chars);
            tokens.push(token);
        }

        Lexer { tokens }
    }

    fn new_helper(char: char, chars: &mut Peekable<impl Iterator<Item = char>>) -> Token {
        match char {
            '+' => Token::Plus,
            ',' => Token::Comma,
            ';' => Token::Semicolon,
            '(' => Token::LParen,
            ')' => Token::RParen,
            '{' => Token::LBrace,
            '}' => Token::RBrace,
            '-' => Token::Minus,
            '*' => Token::Asterisk,
            '/' => Token::Slash,
            '<' => chars.next_if_eq(&'=').map_or(Token::Lt, |_| Token::Lte),
            '>' => chars.next_if_eq(&'=').map_or(Token::Gt, |_| Token::Gte),
            '=' => chars.next_if_eq(&'=').map_or(Token::Assign, |_| Token::Eq),
            '!' => chars.next_if_eq(&'=').map_or(Token::Bang, |_| Token::NotEq),
            '"' => {
                let string: SmolStr = iter::from_fn(|| chars.next_if(|x| *x != '"')).collect();
                chars.next();
                Token::String(string)
            }
            _ if char.is_ascii_alphabetic() || char == '_' => {
                let keyword: SmolStr = iter::once(char)
                    .chain(iter::from_fn(|| {
                        chars.next_if(|x| x.is_ascii_alphanumeric() || *x == '_')
                    }))
                    .collect();

                match keyword.as_str() {
                    "fn" => Token::Function,
                    "let" => Token::Let,
                    "true" => Token::True,
                    "false" => Token::False,
                    "if" => Token::If,
                    "else" => Token::Else,
                    "return" => Token::Return,
                    "nil" => Token::Nil,
                    _ => Token::Identifier(Identifier::new(keyword)),
                }
            }
            _ if char.is_ascii_digit() => iter::once(char)
                .chain(iter::from_fn(|| chars.next_if(|x| x.is_ascii_digit())))
                .collect::<SmolStr>()
                .parse()
                .map_or(Token::Illegal, Token::Int),
            _ => Token::Illegal,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tokenize_sucessfully() {
        let input = r#" 
        let five = 5;
        let ten = 10;

        let add = fn(x, y) {
            x + y;
        };

        let result = add(five, ten);
        !-/*5;
        5 < 10 > 5;

        if (5 < 10) {
            return true;
        } else {
            return false;
        }

        10 == 10;
        10 != 9;
        "foobar"
        "foo bar"
        "#;

        let lexer = Lexer::new(input);

        assert_eq!(
            lexer.tokens,
            vec![
                Token::Let,
                Token::Identifier(SmolStr::new("five").into()),
                Token::Assign,
                Token::Int(5),
                Token::Semicolon,
                Token::Let,
                Token::Identifier(SmolStr::new("ten").into()),
                Token::Assign,
                Token::Int(10),
                Token::Semicolon,
                Token::Let,
                Token::Identifier(SmolStr::new("add").into()),
                Token::Assign,
                Token::Function,
                Token::LParen,
                Token::Identifier(SmolStr::new("x").into()),
                Token::Comma,
                Token::Identifier(SmolStr::new("y").into()),
                Token::RParen,
                Token::LBrace,
                Token::Identifier(SmolStr::new("x").into()),
                Token::Plus,
                Token::Identifier(SmolStr::new("y").into()),
                Token::Semicolon,
                Token::RBrace,
                Token::Semicolon,
                Token::Let,
                Token::Identifier(SmolStr::new("result").into()),
                Token::Assign,
                Token::Identifier(SmolStr::new("add").into()),
                Token::LParen,
                Token::Identifier(SmolStr::new("five").into()),
                Token::Comma,
                Token::Identifier(SmolStr::new("ten").into()),
                Token::RParen,
                Token::Semicolon,
                Token::Bang,
                Token::Minus,
                Token::Slash,
                Token::Asterisk,
                Token::Int(5),
                Token::Semicolon,
                Token::Int(5),
                Token::Lt,
                Token::Int(10),
                Token::Gt,
                Token::Int(5),
                Token::Semicolon,
                Token::If,
                Token::LParen,
                Token::Int(5),
                Token::Lt,
                Token::Int(10),
                Token::RParen,
                Token::LBrace,
                Token::Return,
                Token::True,
                Token::Semicolon,
                Token::RBrace,
                Token::Else,
                Token::LBrace,
                Token::Return,
                Token::False,
                Token::Semicolon,
                Token::RBrace,
                Token::Int(10),
                Token::Eq,
                Token::Int(10),
                Token::Semicolon,
                Token::Int(10),
                Token::NotEq,
                Token::Int(9),
                Token::Semicolon,
                Token::String(SmolStr::new_inline("foobar")),
                Token::String(SmolStr::new_inline("foo bar")),
            ]
        )
    }
}
