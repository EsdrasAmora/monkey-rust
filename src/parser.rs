use self::ast::Statement;
use self::token_parser::TokenParser;
use crate::ast;
use crate::lexer::Lexer;
use crate::token_parser;
use anyhow::Error;

#[derive(Debug)]
pub struct Parser {
    pub nodes: Vec<Statement>,
    pub errors: Vec<Error>,
}

impl Parser {
    pub fn new(lexer: Lexer) -> Self {
        let mut nodes = Vec::new();
        let mut errors = Vec::new();
        let mut tokens = TokenParser::new(lexer.tokens);

        while let Some(current) = tokens.next() {
            tokens
                .parse_statement(current)
                .map_or_else(|err| errors.push(err), |val| nodes.push(val))
        }

        Parser { nodes, errors }
    }
}

//TODO: missing Assign support, only able to declare variables with let but not assign them afterwards;
//TODO: create error types and think about when to buble them up
#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_yaml_snapshot;

    #[test]
    fn parse_call_expression() {
        let input = "
        myFn();
        myFn(1,2);
        myFn(1, 2 * 3, -4 + 5);
        myFn(fn() {});
        fn() {}();
        fn(x) { return x; }(1);
        -myFn(1,2);
        myFn(fn(a,b) {return a + b;}(1,2));";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_fn_expression() {
        let input = "
        fn() {};
        fn(x) { x; };
        fn(x, y, z) { return x + y + z; };";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_grouped_expression() {
        let input = "(5 + 5) * 5; -(5 + 5);";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);

        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_if_expression() {
        let input = "if (x < y) { return x; };";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_if_else_expression() {
        let input = "
        if (x < y) { x; } else { y; };
        if (x < y) { return x; } else { return y; };
        if (x < y) { x == 3; } else { x != 1; };";

        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_infix_expression() {
        let input = "
        5 + 5;
        5 - 5;
        5 * 5;
        5 / 5;
        5 > 5;
        5 < 5;
        5 == 5;
        5 != 5;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);

        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_prefix_expression() {
        let input = "
        -1;
        !5;
        !!-2;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_expression_statement_identifier() {
        let input = "foobar;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_expression_statement_integer_literal() {
        let input = "3;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_let_statement() {
        let input = "
        let x = false;
        let y = 10;
        let foobar = true;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn parse_return_statement() {
        let input = "
        return 5;
        return 10;
        return 993322;";
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }

    #[test]
    fn check_parser_precedence() {
        let input = r#"
        -a * b;
        !-a;
        a + b + c;
        a + b - c;
        a * b * c;
        a * b / c;
        a + b / c;
        a + b * c + d / e - f;
        3 + 4; -5 * 5;
        5 > 4 == 3 < 4;
        5 < 4 != 3 > 4;
        3 + 4 * 5 == 3 * 1 + 4 * 5;
        true;
        false;
        3 > 5 == false;
        3 < 5 == true;
        1 + (2 + 3) + 4;
        (5 + 5) * 2;
        2 / (5 + 5);
        (5 + 5) * 2 * (5 + 5);
        -(5 + 5);
        !(true == true);
        a + add(b * c) + d;
        add(a, b, 1, 2 * 3, 4 + 5, add(6, 7 * 8));
        add(a + b + c * d / f + g);"#;
        let lexer = Lexer::new(input);
        let program = Parser::new(lexer);
        assert!(program.errors.is_empty(), "errors: {:#?}", program.errors);
        assert_yaml_snapshot!(program.nodes);
    }
}
