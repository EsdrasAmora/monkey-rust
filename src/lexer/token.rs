use crate::parser::ast::{BinaryExpression, Expression};
use anyhow::Result;
use serde::{Deserialize, Serialize};
use smol_str::SmolStr;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    Illegal,
    Identifier(Identifier),
    //Literals
    Int(i64),
    String(SmolStr), // a Cow may would be better here.
    True,
    False,
    Nil,
    //Operators
    Eq,
    NotEq,
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    Lt,
    Gt,
    //Delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,
    //Keywords
    Function,
    Let,
    If,
    Else,
    Return,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Identifier(SmolStr);

impl Identifier {
    #[inline]
    pub fn new(name: SmolStr) -> Self {
        Self(name)
    }

    #[inline]
    pub fn into_inner(self) -> SmolStr {
        self.0
    }
}

impl TryFrom<Token> for Identifier {
    type Error = Token;

    #[inline]
    fn try_from(token: Token) -> Result<Self, Self::Error> {
        match token {
            Token::Identifier(name) => Ok(name),
            _ => Err(token),
        }
    }
}

impl From<Identifier> for SmolStr {
    #[inline]
    fn from(name: Identifier) -> Self {
        name.into_inner()
    }
}

impl From<SmolStr> for Identifier {
    #[inline]
    fn from(name: SmolStr) -> Self {
        Self::new(name)
    }
}

impl Token {
    //WTF: why I can't wrap the expressions directly using Some()?
    #[inline]
    pub fn binary_expression_type(&self) -> Option<fn(BinaryExpression) -> Expression> {
        Some(match self {
            Token::Plus => Expression::Add,
            Token::Minus => Expression::Sub,
            Token::Slash => Expression::Div,
            Token::Asterisk => Expression::Mul,
            Token::Eq => Expression::Eq,
            Token::NotEq => Expression::NotEq,
            Token::Lt => Expression::Lt,
            Token::Gt => Expression::Gt,
            _ => return None,
        })
    }

    #[inline]
    pub fn precedence(&self) -> u8 {
        match self {
            Token::Eq | Token::NotEq => 2,
            Token::Gt | Token::Lt => 3,
            Token::Plus | Token::Minus => 4,
            Token::Slash | Token::Asterisk => 5,
            Token::LParen => 7,
            _ => 0,
        }
    }

    #[inline]
    pub fn into_identifier(self) -> Option<Identifier> {
        match self {
            Token::Identifier(name) => Some(name),
            _ => None,
        }
    }

    #[inline]
    pub fn is_identifier(&self) -> bool {
        matches!(self, Token::Identifier(_))
    }

    #[inline]
    pub fn is_literal(&self) -> bool {
        matches!(
            self,
            Token::Int(_)
                | Token::String(_)
                | Token::True
                | Token::False
                | Token::Nil
                | Token::Identifier(_)
        )
    }
}
