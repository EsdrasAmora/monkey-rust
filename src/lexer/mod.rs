pub(crate) mod token;
use std::iter::Peekable;
use token::Token;

#[derive(Debug)]
pub struct Lexer {
    pub tokens: Vec<Token>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        let mut chars = input.trim().chars().filter(|x| x.is_ascii()).peekable();

        let mut temp = String::with_capacity(10);
        let mut tokens = Vec::with_capacity(32);

        while let Some(mut char) = chars.next() {
            while char.is_whitespace() {
                char = chars.next().expect("Unexpected EOF");
            }
            let token = Lexer::new_helper(char, &mut temp, &mut chars);
            tokens.push(token);
        }

        Lexer { tokens }
    }

    fn new_helper(
        char: char,
        temp: &mut String,
        chars: &mut Peekable<impl Iterator<Item = char>>,
    ) -> Token {
        match char {
            '+' => Token::Plus,
            ',' => Token::Comma,
            ';' => Token::Semicolon,
            '(' => Token::LParen,
            ')' => Token::RParen,
            '{' => Token::LBrace,
            '}' => Token::RBrace,
            '-' => Token::Minus,
            '*' => Token::Asterisk,
            '/' => Token::Slash,
            '<' => Token::Lt,
            '>' => Token::Gt,
            '=' => {
                if chars.peek().filter(|x| x == &&'=').is_some() {
                    chars.next();
                    Token::Eq
                } else {
                    Token::Assign
                }
            }
            '!' => {
                if chars.peek().filter(|x| x == &&'=').is_some() {
                    chars.next();
                    Token::NotEq
                } else {
                    Token::Bang
                }
            }
            _ if char.is_ascii_alphabetic() || char == '_' => {
                temp.push(char);

                while let Some(next_char) = chars
                    .peek()
                    .filter(|x| x.is_ascii_alphanumeric() || **x == '_')
                {
                    temp.push(*next_char);
                    chars.next();
                }

                let token = match temp.as_str() {
                    "fn" => Token::Function,
                    "let" => Token::Let,
                    "true" => Token::True,
                    "false" => Token::False,
                    "if" => Token::If,
                    "else" => Token::Else,
                    "return" => Token::Return,
                    _ => Token::Identifier(temp.clone()),
                };

                temp.clear();
                token
            }
            _ if char.is_ascii_digit() => {
                temp.push(char);

                while chars
                    .peek()
                    .filter(|x| x.is_ascii_digit())
                    .map(|x| temp.push(*x))
                    .is_some()
                {
                    chars.next();
                }

                let clone = std::mem::take(temp);
                clone.parse().map_or(Token::Illegal, Token::Int)
            }
            _ => Token::Illegal,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tokenize_sucessfully() {
        let input = r#" 
        let five = 5;
        let ten = 10;

        let add = fn(x, y) {
            x + y;
        };

        let result = add(five, ten);
        !-/*5;
        5 < 10 > 5;

        if (5 < 10) {
            return true;
        } else {
            return false;
        }

        10 == 10;
        10 != 9;"#;

        let lexer = Lexer::new(input);

        assert_eq!(
            lexer.tokens,
            vec![
                Token::Let,
                Token::Identifier("five".to_owned()),
                Token::Assign,
                Token::Int(5),
                Token::Semicolon,
                Token::Let,
                Token::Identifier("ten".to_owned()),
                Token::Assign,
                Token::Int(10),
                Token::Semicolon,
                Token::Let,
                Token::Identifier("add".to_owned()),
                Token::Assign,
                Token::Function,
                Token::LParen,
                Token::Identifier("x".to_owned()),
                Token::Comma,
                Token::Identifier("y".to_owned()),
                Token::RParen,
                Token::LBrace,
                Token::Identifier("x".to_owned()),
                Token::Plus,
                Token::Identifier("y".to_owned()),
                Token::Semicolon,
                Token::RBrace,
                Token::Semicolon,
                Token::Let,
                Token::Identifier("result".to_owned()),
                Token::Assign,
                Token::Identifier("add".to_owned()),
                Token::LParen,
                Token::Identifier("five".to_owned()),
                Token::Comma,
                Token::Identifier("ten".to_owned()),
                Token::RParen,
                Token::Semicolon,
                Token::Bang,
                Token::Minus,
                Token::Slash,
                Token::Asterisk,
                Token::Int(5),
                Token::Semicolon,
                Token::Int(5),
                Token::Lt,
                Token::Int(10),
                Token::Gt,
                Token::Int(5),
                Token::Semicolon,
                Token::If,
                Token::LParen,
                Token::Int(5),
                Token::Lt,
                Token::Int(10),
                Token::RParen,
                Token::LBrace,
                Token::Return,
                Token::True,
                Token::Semicolon,
                Token::RBrace,
                Token::Else,
                Token::LBrace,
                Token::Return,
                Token::False,
                Token::Semicolon,
                Token::RBrace,
                Token::Int(10),
                Token::Eq,
                Token::Int(10),
                Token::Semicolon,
                Token::Int(10),
                Token::NotEq,
                Token::Int(9),
                Token::Semicolon
            ]
        )
    }
}
