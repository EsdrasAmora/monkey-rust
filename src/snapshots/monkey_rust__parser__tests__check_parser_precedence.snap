---
source: src/parser.rs
expression: program.nodes
---
- Expression:
    BinaryExp:
      operator: Mul
      lhs:
        UnaryExpression:
          value:
            Identifier: a
          operator: Oposite
      rhs:
        Identifier: b
- Expression:
    UnaryExpression:
      value:
        UnaryExpression:
          value:
            Identifier: a
          operator: Oposite
      operator: Not
- Expression:
    BinaryExp:
      operator: Add
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            Identifier: a
          rhs:
            Identifier: b
      rhs:
        Identifier: c
- Expression:
    BinaryExp:
      operator: Sub
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            Identifier: a
          rhs:
            Identifier: b
      rhs:
        Identifier: c
- Expression:
    BinaryExp:
      operator: Mul
      lhs:
        BinaryExp:
          operator: Mul
          lhs:
            Identifier: a
          rhs:
            Identifier: b
      rhs:
        Identifier: c
- Expression:
    BinaryExp:
      operator: Div
      lhs:
        BinaryExp:
          operator: Mul
          lhs:
            Identifier: a
          rhs:
            Identifier: b
      rhs:
        Identifier: c
- Expression:
    BinaryExp:
      operator: Add
      lhs:
        Identifier: a
      rhs:
        BinaryExp:
          operator: Div
          lhs:
            Identifier: b
          rhs:
            Identifier: c
- Expression:
    BinaryExp:
      operator: Sub
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            BinaryExp:
              operator: Add
              lhs:
                Identifier: a
              rhs:
                BinaryExp:
                  operator: Mul
                  lhs:
                    Identifier: b
                  rhs:
                    Identifier: c
          rhs:
            BinaryExp:
              operator: Div
              lhs:
                Identifier: d
              rhs:
                Identifier: e
      rhs:
        Identifier: f
- Expression:
    BinaryExp:
      operator: Add
      lhs:
        Literal:
          Int: 3
      rhs:
        Literal:
          Int: 4
- Expression:
    BinaryExp:
      operator: Mul
      lhs:
        UnaryExpression:
          value:
            Literal:
              Int: 5
          operator: Oposite
      rhs:
        Literal:
          Int: 5
- Expression:
    BinaryExp:
      operator: Eq
      lhs:
        BinaryExp:
          operator: Gt
          lhs:
            Literal:
              Int: 5
          rhs:
            Literal:
              Int: 4
      rhs:
        BinaryExp:
          operator: Lt
          lhs:
            Literal:
              Int: 3
          rhs:
            Literal:
              Int: 4
- Expression:
    BinaryExp:
      operator: NotEq
      lhs:
        BinaryExp:
          operator: Lt
          lhs:
            Literal:
              Int: 5
          rhs:
            Literal:
              Int: 4
      rhs:
        BinaryExp:
          operator: Gt
          lhs:
            Literal:
              Int: 3
          rhs:
            Literal:
              Int: 4
- Expression:
    BinaryExp:
      operator: Eq
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            Literal:
              Int: 3
          rhs:
            BinaryExp:
              operator: Mul
              lhs:
                Literal:
                  Int: 4
              rhs:
                Literal:
                  Int: 5
      rhs:
        BinaryExp:
          operator: Add
          lhs:
            BinaryExp:
              operator: Mul
              lhs:
                Literal:
                  Int: 3
              rhs:
                Literal:
                  Int: 1
          rhs:
            BinaryExp:
              operator: Mul
              lhs:
                Literal:
                  Int: 4
              rhs:
                Literal:
                  Int: 5
- Expression:
    Literal: "True"
- Expression:
    Literal: "False"
- Expression:
    BinaryExp:
      operator: Eq
      lhs:
        BinaryExp:
          operator: Gt
          lhs:
            Literal:
              Int: 3
          rhs:
            Literal:
              Int: 5
      rhs:
        Literal: "False"
- Expression:
    BinaryExp:
      operator: Eq
      lhs:
        BinaryExp:
          operator: Lt
          lhs:
            Literal:
              Int: 3
          rhs:
            Literal:
              Int: 5
      rhs:
        Literal: "True"
- Expression:
    BinaryExp:
      operator: Add
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            Literal:
              Int: 1
          rhs:
            BinaryExp:
              operator: Add
              lhs:
                Literal:
                  Int: 2
              rhs:
                Literal:
                  Int: 3
      rhs:
        Literal:
          Int: 4
- Expression:
    BinaryExp:
      operator: Mul
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            Literal:
              Int: 5
          rhs:
            Literal:
              Int: 5
      rhs:
        Literal:
          Int: 2
- Expression:
    BinaryExp:
      operator: Div
      lhs:
        Literal:
          Int: 2
      rhs:
        BinaryExp:
          operator: Add
          lhs:
            Literal:
              Int: 5
          rhs:
            Literal:
              Int: 5
- Expression:
    BinaryExp:
      operator: Mul
      lhs:
        BinaryExp:
          operator: Mul
          lhs:
            BinaryExp:
              operator: Add
              lhs:
                Literal:
                  Int: 5
              rhs:
                Literal:
                  Int: 5
          rhs:
            Literal:
              Int: 2
      rhs:
        BinaryExp:
          operator: Add
          lhs:
            Literal:
              Int: 5
          rhs:
            Literal:
              Int: 5
- Expression:
    UnaryExpression:
      value:
        BinaryExp:
          operator: Add
          lhs:
            Literal:
              Int: 5
          rhs:
            Literal:
              Int: 5
      operator: Oposite
- Expression:
    UnaryExpression:
      value:
        BinaryExp:
          operator: Eq
          lhs:
            Literal: "True"
          rhs:
            Literal: "True"
      operator: Not
- Expression:
    BinaryExp:
      operator: Add
      lhs:
        BinaryExp:
          operator: Add
          lhs:
            Identifier: a
          rhs:
            Call:
              arguments:
                - BinaryExp:
                    operator: Mul
                    lhs:
                      Identifier: b
                    rhs:
                      Identifier: c
              function:
                Identifier: add
      rhs:
        Identifier: d
- Expression:
    Call:
      arguments:
        - Identifier: a
        - Identifier: b
        - Literal:
            Int: 1
        - BinaryExp:
            operator: Mul
            lhs:
              Literal:
                Int: 2
            rhs:
              Literal:
                Int: 3
        - BinaryExp:
            operator: Add
            lhs:
              Literal:
                Int: 4
            rhs:
              Literal:
                Int: 5
        - Call:
            arguments:
              - Literal:
                  Int: 6
              - BinaryExp:
                  operator: Mul
                  lhs:
                    Literal:
                      Int: 7
                  rhs:
                    Literal:
                      Int: 8
            function:
              Identifier: add
      function:
        Identifier: add
- Expression:
    Call:
      arguments:
        - BinaryExp:
            operator: Add
            lhs:
              BinaryExp:
                operator: Add
                lhs:
                  BinaryExp:
                    operator: Add
                    lhs:
                      Identifier: a
                    rhs:
                      Identifier: b
                rhs:
                  BinaryExp:
                    operator: Div
                    lhs:
                      BinaryExp:
                        operator: Mul
                        lhs:
                          Identifier: c
                        rhs:
                          Identifier: d
                    rhs:
                      Identifier: f
            rhs:
              Identifier: g
      function:
        Identifier: add

